Param rework

lalligagger opened the pull request:
@ivandorte I left a bunch of params still to be filled in, but I think you will know better how to set many of them up w/ plotting. Basically:

- Brings back MapManager parameterized class to handle data/ plot state
- Chat/ agent now has history
- Attempted to generate GRIBS basemap tile url, but not working yet
- Removed temporary rgb plot function, using image_plots module again

Todos as soon as we can, at which point I'm comfortable cutting our video and submitting for the contest if you are:

- Get spyndex and cloud mask selections working for Landsat/ in chatbox
- Finalize other (param) controls that will be attached to datacube
- Finalize any other plotting features you're ready to
- Take one more crack at MODIS/GRIBS basemaps

We can track in new issues but wanted to give you the main punch list here.

lalligagger committed:
tools update

lalligagger committed:
comments

lalligagger committed:
started reworks/ cleanup

lalligagger committed:
started reworks/ cleanup

lalligagger committed:
started reworks/ cleanup

lalligagger committed:
everything to methods, chat to ascending

lalligagger committed:
black/ flake

lalligagger committed:
use img_proc

lalligagger committed:
formatting

lalligagger committed:
formatting

lalligagger committed:
chatbox now uses map_mgr media attrib

lalligagger committed:
landsat search working, spyndex constants added

lalligagger committed:
switched to image_plot true color

ivandorte commented on modules/chat_utils.py:
```diff
@@ -0,0 +1,194 @@
+        layer = f"MODIS_{source}_CorrectedReflectance_TrueColor"
+        base_url = "https://gibs.earthdata.nasa.gov/wmts/epsg3857/best/"
+        tile_path = f"{layer}/default/{datestring}/{tileMatrixSet}/" + "{Z}/{Y}/{X}.jpg"
+        self.tile_url = base_url + tile_path
```
The url you constructed works with hv.Tiles(tile_url). However, to prevent a blank basemap you need to add the max_zoom option:

hv.Tiles(tile_url).opts(max_zoom=9)

ivandorte commented on modules/spyndex_utils.py:
```diff
@@ -16,6 +16,20 @@
     "S2": "swir22",
 }
 
+LANDSAT_BAND_MAPPING = {
```
I think we don't need this dictionary. We can use a generic dictionary (e.g. BAND_MAPPING) where we can add all the bands of the platforms we will use.

ivandorte committed:
Added a function to get landsat indices

Signed-off-by: ivandorte <edgar_allan_poe88@hotmail.it>

lalligagger commented:
> I think we don't need this dictionary. We can use a generic dictionary (e.g. BAND_MAPPING) where we can add all the bands of the platforms we will use.

The issue I saw here was even for Landsat vs. Sentinel we have e.g. "nir" vs. "nir08" for what I assume would be "N" in both cases? Maybe this is already handled in spyndex?

lalligagger committed:
added rgb+indices widget

lalligagger committed:
added cmask widget

lalligagger committed:
range param added & bound to contrast stretch

lalligagger committed:
cmask select working for s2

lalligagger committed:
landsat cmask working but there are scaling issues

lalligagger commented:
I believe the scaling issues are because Landsat wants an offset & scaling where Sentinel just does scaling (implemented in your `s2_dn_to_reflectance`).

Here is what I'm thinking to close this out:

1. There should be a `_load_data` function which updates the `data` attribute of the map_mgr
2. This will look at map_mgr.collection, apply the correct L2 scaling function, decide how to apply cloud mask, and map any band names as needed
3. The plot functions (RGB or spyndex) will take in data that has already been processed per 2

Will start working on this now but lmk what you think.

lalligagger committed:
comments on _load_data updates

lalligagger commented:
Hmmm... feels like i am close but this isn't working for me

the plot triggers but the data seems to stay the same. looking into it.

![image](https://github.com/lalligagger/satgpt-app/assets/11700267/4b3e1817-6520-49fc-8973-ad2f1e711a44)

ivandorte commented:
> > I think we don't need this dictionary. We can use a generic dictionary (e.g. BAND_MAPPING) where we can add all the bands of the platforms we will use.
> 
> The issue I saw here was even for Landsat vs. Sentinel we have e.g. "nir" vs. "nir08" for what I assume would be "N" in both cases? Maybe this is already handled in spyndex?

Yeah, you are right! Better to keep two dictionaries. BAND_MAPPING -> S2_BAND_MAPPING.

ivandorte commented:
> I believe the scaling issues are because Landsat wants an offset & scaling where Sentinel just does scaling (implemented in your `s2_dn_to_reflectance`).
> 
> Here is what I'm thinking to close this out:
> 
> 1. There should be a `_load_data` function which updates the `data` attribute of the map_mgr
> 2. This will look at map_mgr.collection, apply the correct L2 scaling function, decide how to apply cloud mask, and map any band names as needed
> 3. The plot functions (RGB or spyndex) will take in data that has already been processed per 2
> 
> Will start working on this now but lmk what you think.

I was working on something similar for load_data, because for other plots I need all the bands.
In the map manager i request the s2_data with all bands and all items. Then in the true color plot function i filter the s2_data by bands (in this case r, g, b) and time (need to convert the time into string to make it work, no need for hms because xarray select the nearest date).
For Landsat, we can't apply the dn_to_reflectance. We can't divide by 1e4, maybe we need to divide by 65563 and clip to 1 (min) - 65563 (max).

ivandorte commented:
I just pass all the items to stac_load in the map manager and store to a variable (e.g sat_data). This will be the input for the true color binding. I have removed items in favour of a generic variable e.g "sat_data" or "data".

ivandorte commented:
> Hmmm... feels like i am close but this isn't working for me (all under `map_mgr._viewer`)
> 
> the plot triggers but the data seems to stay the same. looking into it.
> 
> ![image](https://github.com/lalligagger/satgpt-app/assets/11700267/4b3e1817-6520-49fc-8973-ad2f1e711a44)
> 

I just pass all the items to stac_load in the map manager and store to a variable (e.g sat_data). This will be the input for the true color binding. I have removed items in favour of a generic variable e.g "sat_data" or "data".

lalligagger commented:
I am about to toss this over to you. I had the cloud masks working, but not the time selection. Now I got the time selection working & cleaned things up, but cloud mask doesn't seem to work.

You found the right link on scaling for Landsat L2! One equation each for OLI and TIRS.

lalligagger committed:
time selection working, but cmask broke

lalligagger committed:
added bbox to _load_data

lalligagger commented:
> I just pass all the items to stac_load in the map manager and store to a variable (e.g sat_data). This will be the input for the true color binding. I have removed items in favour of a generic variable e.g "sat_data" or "data".

We are definitely on the same page here! (See last commit, cleaned up significantly.) Just the cloud mask black nodata issue on the RGB tests I was running. Landsat working up to proper scaling.

Hopefully it is easy for you to work in your latest plotting functions from here! I can hit merge before I go to bed, just lmk if that's helpful.

ivandorte commented:
> > I just pass all the items to stac_load in the map manager and store to a variable (e.g sat_data). This will be the input for the true color binding. I have removed items in favour of a generic variable e.g "sat_data" or "data".
> 
> We are definitely on the same page here! (See last commit, cleaned up significantly.) Just the cloud mask black nodata issue on the RGB tests I was running. Landsat working up to proper scaling.
> 
> Hopefully it is easy for you to work in your latest plotting functions from here! I can hit merge before I go to bed, just lmk if that's helpful.

Yes, for me you can merge, I will create another branch and try to add the plotting function to the datacube.
